### ISSUE TITLE
AMOEBA RNA forcefield cannot match bonding arrangement in PDB

### ISSUE THREAD

**Issue by lvotapka** [openmm/openmm#3305]
Hi @peastman,

Great job finishing the 2018 AMOEBA, and I also just wanted to say thanks a million for including the nucleic acids in there.

I'm running into this error with the following inputs:

ValueError: No template found for residue 1 (RU5).  The set of atoms matches RU5, but the bonds are different.  Perhaps the chain is missing a terminal group?

This one was built using the sequence command in LEAP. I've also had the same error when I use NAB and then loadpdb in LEAP to make an input PDB. The chain should not be missing a terminal group, because the 5'  end does indeed have the 5' uracil cap on this end (resname RU5).

[rna_ext.pdb.zip](https://github.com/openmm/openmm/files/7437108/rna_ext.pdb.zip)

By the way, I'm using
`forcefield = ForceField('amoeba2018.xml')`

**Comment by jchodera** [comment]
This is presumably occurring because your PDB file uses only `ATOM` cards without `CONECT` cards to specify bonds, so OpenMM is forced to use [these rules](https://github.com/openmm/openmm/blob/master/wrappers/python/openmm/app/data/pdbNames.xml) to fill in bond information for any standard residues specified in the [PDB format specification](https://www.wwpdb.org/documentation/file-format). The standard residues listed in the spec are:
![image](https://user-images.githubusercontent.com/3656088/139553892-31da1a15-0e08-4072-9d32-7a137107e97b.png)
Because your PDB file uses a nonstandard naming convention (`RU`, `RU5` are not standard names), the OpenMM [`createStandardResidues`](https://github.com/openmm/openmm/blob/master/wrappers/python/openmm/app/topology.py#L296-L342) method doesn't know which standard bonds to add since it assumes you have used the PDB standard residue names when writing a PDB file.

*TL;DR:* Your PDB file does not conform to the PDB standard, so you either need to rename the residues or ensure you write `CONECT` records. If you prepared this in LEaP, you should be able to use `ambpdb` with the `-conect` flag to convert the resulting prmtop/inpcrd files to a PDB file that contains `CONECT` records.

**Comment by jchodera** [comment]
@peastman: Maybe we want to add the `ambdpdb -conect` suggestion to the [FAQ](https://github.com/openmm/openmm/wiki/Frequently-Asked-Questions#why-am-i-getting-an-error-that-says-no-template-found-for-residue)?

**Comment by lvotapka** [comment]
Hi @jchodera, thanks for the helpful suggestion, indeed, using the -conect argument in ambpdb seemed to get us past this error. However, now a new one has arose: 

  File "/home/lvotapka/miniconda3/lib/python3.8/site-packages/openmm/app/forcefield.py", line 5445, in createForce
    raise ValueError("Atom %s of %s %d was not assigned." %(atom.name, atom.residue.name, atom.residue.index))
ValueError: Atom O2' of U 0 was not assigned.

I'm attaching a new PDB file with the CONECT keywords intact, and you can see that they are named according to the standard residues specified in the link you sent. It also appears that O2' exists in the PDB file, as well as within the pdbNames.xml file.

[rna_ext_conect.pdb.zip](https://github.com/openmm/openmm/files/7454013/rna_ext_conect.pdb.zip)

Here is a script I'm using to run the simulation, in case that would be helpful:

`from openmm.app import *
from openmm import *
from openmm.unit import *
from sys import stdout
import time

cuda_device_index = "0"
num_steps=10000
timestep = 0.002*picoseconds

pdb = PDBFile("rna_ext_conect.pdb")
forcefield = ForceField("amoeba2018.xml")
system = forcefield.createSystem(pdb.topology, nonbondedMethod=PME, nonbondedCutoff=1*nanometer, constraints=HBonds)
integrator = LangevinIntegrator(300*kelvin, 1/picosecond, timestep)
platform = Platform.getPlatformByName("CUDA")
properties = {"CudaDeviceIndex": cuda_device_index, "CudaPrecision": "mixed"}
simulation = Simulation(pdb.topology, system, integrator, platform, properties)
simulation.context.setPositions(pdb.positions)
simulation.minimizeEnergy()
simulation.reporters.append(PDBReporter("output.pdb", 1000))
simulation.reporters.append(StateDataReporter(stdout, 1000, step=True, potentialEnergy=True, temperature=True))

start_time = time.time()
simulation.step(num_steps)
total_time = time.time() - start_time
simulation_in_ns = num_steps * timestep.value_in_unit(picoseconds) * 1e-3
total_time_in_days = total_time / (86400.0)
ns_per_day = simulation_in_ns / total_time_in_days
print("Equilibration benchmark:", ns_per_day, "ns/day")`

**Comment by peastman** [comment]
I'm not sure why your file doesn't work, but I can't get it to work with Tinker either.  I first converted it to an xyz file with

```
pdbxyz rna_ext_conect.pdb
```

and then tried to compute energy with

```
analyze rna_ext_conect.xyz
```

Tinker prints out a huge list of errors, starting with

```
 Atoms with an Unusual Number of Attached Atoms :

 Type           Atom Name      Atom Type       Expected    Found

 Valence           2-C             82              3         2
 Valence           3-C             64              3         4
 Valence           6-C             64              3         4
 Valence           8-C             82              3         2
```

and ending with

```
 MECHANIC  --  Some Required Potential Energy Parameters are Undefined

 Tinker is Unable to Continue; Terminating the Current Calculation
```

I tried removing the CONECT records from your PDB file, but it still gives the same errors.

**Comment by lvotapka** [comment]
Oh wow, that is indeed very strange. Thanks for going to so much trouble to get it to work @peastman. I don't feel tied down to this particular PDB, so how did you generate your nucleic acid structures for tests, etc.?

**Comment by peastman** [comment]
The one I used in the test case, I just loaded a PDB file (1A1F) that happened to include a short DNA into PDBFixer and had it delete everything except that chain.

**Comment by lvotapka** [comment]
Yes, I am also trying to figure out what is wrong with it. I'll keep this thread updated as I find things. FYI, the structures were made using NAB and LEAP in Ambertools. I wonder if the atom naming/ordering is just different in Ambertools than it would be in Tinker??

One thought, @peastman, are you sure that the pdbNames.xml file is correctly distinguishing between DNA and RNA? I'm noticing that the problems seem to stem from the existence of the O2' atom in the RNA, which is the one structural difference between DNA and RNA. I see that O2' exists in the "Nucleic" Residue category, but that there doesn't seem to be any distinguishing differences between G and DG, and C and DC, for instance. I don't fully understand how pdbNames.xml works, but it's just a thought.

**Comment by peastman** [comment]
pdbNames.xml lists common, nonstandard names that it should recognize.  In this case the line is

```xml
  <Atom name="O2'" alt1="O2*"/>
```

That tells it that if it sees an atom called "O2*" in a nucleic acid, it should change the name to "O2'".

**Comment by lvotapka** [comment]
Ah OK gotcha.

**Comment by lvotapka** [comment]
I'm planning to find another test case on the PDB to use for RNA to see if I can get an RNA molecule to work at least.

**Comment by lvotapka** [comment]
Aha! I attempted to run PBDID: 1HS2 through PDBfixer and then run with AMOEBA and I got essentially the same error:

raise ValueError("Atom %s of %s %d was not assigned." %(atom.name, atom.residue.name, atom.residue.index))
ValueError: Atom O2' of G 0 was not assigned.

**Comment by peastman** [comment]
Thanks, that was a perfect test case.  The fix is in #3313.

**Comment by lvotapka** [comment]
Glad I could help, and thank YOU, @peastman, you're the programmer I aspire to be!
